### PR TITLE
Appman 2576 access change history all learner records

### PR DESCRIPTION
**Pull request by skumardfe**

### CODE REVIEW

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Mappers/Apprentice/IndexViewModelMapper.cs [hunk 1]
```diff
@@ -78,8 +77,8 @@ public async Task<IndexViewModel> Map(IndexRequest source)
 
         if (response.TotalApprenticeships >= Constants.ApprenticesSearch.NumberOfApprenticesRequiredForSearch)
         {
-            var filters = await _client.GetApprenticeshipsFilterValues(
-                new GetApprenticeshipFiltersRequest{ProviderId = source.ProviderId});
+            var filters = await _outerApiService.GetApprenticeshipsFilters(
```
Comment (cofaulco) [inline src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Mappers/Apprentice/IndexViewModelMapper.cs]: there should only be a single OuterApi call in this mapper. Return the filters data on the GetApprentices response

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments/Infrastructure/OuterApi/IOuterApiService.cs [hunk 2]
```diff
@@ -18,45 +18,83 @@ namespace SFA.DAS.ProviderCommitments.Interfaces;
...
+
+    Task<SyncLearnerDataResponse> SyncLearnerData(long providerId, long cohortId, long draftApprenticeshipId);
+
     Task DraftApprenticeshipSetReference(long providerId, long cohortId, long apprenticeshipId, DraftApprenticeshipSetReferenceApimRequest request);
+
     Task DraftApprenticeshipAddEmail(long providerId, long cohortId, long apprenticeshipId, DraftApprenticeAddEmailApimRequest request);
+
+    Task<GetApprenticeshipsResponse> GetApprenticeships(GetApprenticeshipsRequest request);
+
+    Task<GetApprenticeshipsFiltersResponse> GetApprenticeshipsFilters(GetApprenticeshipsFiltersRequest request);
```
Comment (cofaulco) [inline src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments/Infrastructure/OuterApi/IOuterApiService.cs]: Remove implementation of GetApprenticeshipsFilters

**Review by cofaulco — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Mappers/Apprentice/IndexViewModelMapper.cs (hunk 1 above)
Comment (skumardfe) [reply src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Mappers/Apprentice/IndexViewModelMapper.cs]: done

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments/Infrastructure/OuterApi/IOuterApiService.cs (hunk 2 above)
Comment (skumardfe) [reply src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments/Infrastructure/OuterApi/IOuterApiService.cs]: done

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Controllers/ApprenticeController.cs [hunk 3]
```diff
@@ -723,6 +723,14 @@ public async Task<IActionResult> ChangeHistory(ChangeHistoryRequest request)
         return View(viewModel);
     }
 
+    [Route("get-all-change-history")]
```
Comment (cofaulco) [inline src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Controllers/ApprenticeController.cs]: this should just be `change-history`

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Mappers/Apprentice/GetAllChangeHistoryRequestToViewModelMapper.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+            ChangeHistory = changeHistory.ChangeHistory.ConvertAll(x => new GetAllChangeHistoryViewModel
+            {
+                Description = x.Description,
+                AppliedDate = x.AppliedDate,
+                ChangeType = (LearningChangeType)x.ChangeType,
+                Id = x.Id,
+                LearnerName = x.LearnerName,
+                EmployerName = x.EmployerName
+            }),
+            AvailableFrom = changeHistory.ChangeHistory.OrderBy(t => t.Created).FirstOrDefault()?.Created ?? DateTime.UtcNow
```
Comment (cofaulco) [inline src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Mappers/Apprentice/GetAllChangeHistoryRequestToViewModelMapper.cs:26]: changeHistory and changeHistory.ChangeHistory are dereferenced without null checks. If the API returns null or an empty body, this throws. Guard like the BFF handler (changeHistory?.ChangeHistory ?? []) and add a test for the empty/null case.

**Review by cofaulco — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Controllers/ApprenticeController.cs (hunk 3 above)
Comment (skumardfe) [reply src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Controllers/ApprenticeController.cs]: done.

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Mappers/Apprentice/GetAllChangeHistoryRequestToViewModelMapper.cs (hunk 4 above)
Comment (skumardfe) [reply src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Mappers/Apprentice/GetAllChangeHistoryRequestToViewModelMapper.cs:26]: approvals is sending empty and a new test case is added for empty result.

**Review by cofaulco — DISMISSED** [review]

**Review by Pauljgraham — APPROVED** [review]
